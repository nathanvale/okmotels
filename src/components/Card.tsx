import React from 'react'
import styled from 'styled-components'
import {Link} from 'gatsby'
import Img from 'gatsby-image'
import {Maybe, ContentfulAsset, MarkdownRemark} from '../graphql'

import {Fluid} from '../types'

//extends Pick<ContentfulPost, 'title' | 'id' | 'slug' | 'publishDate'>
interface Props {
  featured?: boolean
  // heroImage: Maybe<Pick<ContentfulAsset, 'title'> & {fluid: Fluid}>
  // body: Maybe<{
  //   childMarkdownRemark: Maybe<Pick<MarkdownRemark, 'html' | 'excerpt'>>
  // }>
}

const Post = styled.li<Props>`
  position: relative;
  border: 1px solid ${props => props.theme.colors.secondary};
  border-radius: 2px;
  margin: 0 0 1em 0;
  width: 100%;
  transition: background 0.2s;
  @media screen and (min-width: ${props => props.theme.responsive.small}) {
    flex: ${props => (props.featured ? '0 0 100%' : '0 0 49%')};
    margin: 0 0 2vw 0;
  }
  @media screen and (min-width: ${props => props.theme.responsive.medium}) {
    flex: ${props => (props.featured ? '0 0 100%' : '0 0 32%')};
  }
  &:hover {
    background: ${props => props.theme.colors.tertiary};
  }
  a {
    display: flex;
    flex-flow: column;
    height: 100%;
    width: 100%;
    color: ${props => props.theme.colors.base};
    text-decoration: none;
    .gatsby-image-wrapper {
      height: 0;
      padding-bottom: 60%;
      @media screen and (min-width: ${props => props.theme.responsive.small}) {
        padding-bottom: ${props => (props.featured ? '40%' : '60%')};
      }
    }
  }
`

const Title = styled.h2`
  font-size: 1.5em;
  font-weight: 600;
  text-transform: capitalize;
  margin: 1rem 1rem 0.5rem 1rem;
`

const Date = styled.h3`
  margin: 0 1rem 1.5rem 1rem;
  color: gray;
`

const Excerpt = styled.p`
  margin: 0 1rem 1rem 1rem;
  line-height: 1.6;
`

const Card: React.SFC<Props> = ({
  slug,
  heroImage,
  title,
  publishDate,
  body,
  ...props
}): JSX.Element => {
  return (
    <Post featured={props.featured}>
      <Link to={`/${slug}/`}>
        <Img fluid={heroImage && heroImage.fluid} backgroundColor="#eeeeee" />
        <Title>{title}</Title>
        <Date>{publishDate}</Date>
        <Excerpt
          dangerouslySetInnerHTML={{
            __html:
              body &&
              body.childMarkdownRemark &&
              body.childMarkdownRemark.excerpt,
          }}
        />
      </Link>
    </Post>
  )
}

export default Card
