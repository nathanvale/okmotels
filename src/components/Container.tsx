import React from 'react'
import styled from 'styled-components'

interface ContainerProps {
  ['data-testid']: string
}

const Wrapper = styled.section`
  margin: 0 auto auto;
  width: 100%;
  max-width: ${props => props.theme.sizes.maxWidth};
  padding: 3em 1.5em 2em;
  flex-grow: 1;
`

const Container: React.FC<ContainerProps> = ({
  'data-testid': dataTestId,
  children,
}): JSX.Element => {
  return <Wrapper data-testid={dataTestId}>{children}</Wrapper>
}

export {Container}
