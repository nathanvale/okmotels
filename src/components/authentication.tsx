import styled, {css} from 'styled-components'
import React from 'react'
import {useAuth, AuthContextObject} from '../context/auth-context'
import {useCallbackStatus} from '../utils/use-callback-status'
import {Modal} from './Modal'
import {Button, FormGroup, Spinner} from './lib'

interface UserFormProps {
  handleSubmit(event: React.FormEvent<HTMLFormElement>): void
  buttonText: string
  isPending: boolean
  error: Error | null
  isRejected: boolean
}

interface UserPayload {
  username: string
  password: string
}

interface VerifyUserFormProps {
  verifyEmail: AuthContextObject['verifyEmail']
  email: string
  onVerified: () => void
}

const ModalTitle = styled.h3({
  textAlign: 'center',
  fontSize: '2em',
})

export function LoginForm() {
  const {login, verifyEmail} = useAuth()
  const {isPending, isRejected, error, run} = useCallbackStatus()
  const [isVerifyRequired, setIsVerifyRequired] = React.useState(false)
  const [user, setUser] = React.useState<UserPayload | undefined>()
  function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault()
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const {username, password} = (event.target as any).elements

    const payload: UserPayload = {
      username: username.value,
      password: password.value,
    }

    setUser(payload)

    run(
      login(payload).catch((e: Error) => {
        if (e.message === 'User is not confirmed.') {
          setIsVerifyRequired(true)
        } else {
          throw e
        }
      }),
    )
  }

  function onVerified() {
    setIsVerifyRequired(false)
    if (user) run(login(user))
  }

  return isVerifyRequired && user ? (
    <VerifyUserForm
      email={user.username}
      verifyEmail={verifyEmail}
      onVerified={onVerified}
    />
  ) : (
    <UserForm
      buttonText="Login"
      handleSubmit={handleSubmit}
      isPending={isPending}
      error={error}
      isRejected={isRejected}
    />
  )
}

function RegisterForm() {
  const {login, register, verifyEmail} = useAuth()
  const [isUserRegistered, setIsUserRegistered] = React.useState(false)
  const [isUserVerified, setIsUserVerified] = React.useState(false)
  const [user, setUser] = React.useState<UserPayload | undefined>()
  const {isPending, isRejected, error, run} = useCallbackStatus()

  function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault()
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const {username, password} = (event.target as any).elements
    const payload: UserPayload = {
      username: username.value,
      password: password.value,
    }

    setUser(payload)

    run(
      register(payload).then(() => {
        setIsUserRegistered(true)
      }),
    )
  }

  function onVerified() {
    setIsUserVerified(true)
    if (user) run(login(user))
  }

  return isUserRegistered && !isUserVerified && user ? (
    <VerifyUserForm
      email={user.username}
      verifyEmail={verifyEmail}
      onVerified={onVerified}
    />
  ) : (
    <UserForm
      buttonText="Register"
      handleSubmit={handleSubmit}
      isPending={isPending}
      error={error}
      isRejected={isRejected}
    />
  )
}

function UserForm({
  handleSubmit,
  buttonText,
  isPending,
  isRejected,
  error,
}: UserFormProps) {
  return (
    <form
      onSubmit={handleSubmit}
      css={css({
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'stretch',
        '> div': {
          margin: '10px auto',
          width: '100%',
          maxWidth: '300px',
        },
      })}
    >
      <FormGroup>
        <label htmlFor="username">Username</label>
        <input id="username" />
      </FormGroup>
      <FormGroup>
        <label htmlFor="password">Password</label>
        <input id="password" type="password" />
      </FormGroup>
      <div>
        <Button type="submit">
          {buttonText}{' '}
          {isPending ? <Spinner css={css({marginLeft: 5})} /> : null}
        </Button>
      </div>
      {isRejected ? (
        <div css={css({color: 'red'})}>{error ? error.message : null}</div>
      ) : null}
    </form>
  )
}

function VerifyUserForm({verifyEmail, email, onVerified}: VerifyUserFormProps) {
  const {isPending, isRejected, error, run} = useCallbackStatus()

  function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault()
    const {verificationCode} = (event.target as any).elements

    run(
      verifyEmail({
        email,
        verificationCode: verificationCode.value,
      }).then(onVerified),
    )
  }
  return (
    <form
      onSubmit={handleSubmit}
      css={css({
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'stretch',
        '> div': {
          margin: '10px auto',
          width: '100%',
          maxWidth: '300px',
        },
      })}
    >
      <FormGroup>
        <label htmlFor="verificationCode">Verification Code</label>
        <input id="verificationCode" />
      </FormGroup>
      <div>
        Enter the verifaction code that was sent to the email address you
        provided
        <Button type="submit">
          Verify email address{' '}
          {isPending ? <Spinner css={css({marginLeft: 5})} /> : null}
        </Button>
      </div>
      {isRejected ? (
        <div css={css({color: 'red'})}>{error ? error.message : null}</div>
      ) : null}
    </form>
  )
}

interface LoginModalProps {
  trigger: React.ReactElement
}

interface RegisterModalProps {
  trigger: React.ReactElement
}

export function LoginModal({trigger}: LoginModalProps) {
  return (
    <Modal trigger={trigger}>
      <ModalTitle>Login</ModalTitle>
      <LoginForm />
    </Modal>
  )
}

export function RegisterModal({trigger}: RegisterModalProps) {
  return (
    <Modal trigger={trigger}>
      <ModalTitle>Register</ModalTitle>
      <RegisterForm />
    </Modal>
  )
}
