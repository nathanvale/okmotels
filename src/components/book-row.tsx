import React from 'react'
import {Link} from '@reach/router'
import {DefaultTheme, css} from 'styled-components'
import {useSingleListItemState} from '../context/list-item-context'
import {Book} from '../../types/custom-types'
import {media} from '../styles/media-queries'
import {StatusButtons} from './status-buttons'
import {Rating} from './rating'

interface BookRowProps {
  book: Book
  theme?: DefaultTheme
}

function BookRow({book, theme}: BookRowProps) {
  const {title, author, coverImageUrl} = book
  const listItem = useSingleListItemState(book.id)
  return (
    <div
      css={css({
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'flex-end',
        position: 'relative',
      })}
    >
      <Link
        to={`/app/book/${book.id}`}
        css={css({
          flexGrow: 2,
          display: 'grid',
          gridTemplateColumns: '140px 1fr',
          gridGap: 20,
          border: `1px solid ${theme && theme.colors.gray20}`,
          color: theme && theme.colors.text,
          padding: '1.25em',
          borderRadius: '3px',
          ':hover,:focus': {
            textDecoration: 'none',
            boxShadow: '0 5px 15px -5px rgba(0,0,0,.08)',
            color: 'inherit',
          },
        })}
      >
        <div
          css={css({
            width: 140,
            [media.sm]: {
              width: 100,
            },
          })}
        >
          <img
            src={coverImageUrl}
            alt={`${title} book cover`}
            css={css({maxHeight: '100%', maxWidth: '100%'})}
          />
        </div>
        <div css={css({flex: 1})}>
          <div css={css({display: 'flex', justifyContent: 'space-between'})}>
            <div css={css({flex: 1})}>
              <h2
                css={css({
                  fontSize: '1.25em',
                  margin: '0',
                  color: theme && theme.colors.indigo,
                })}
              >
                {title}
              </h2>
              {listItem ? <Rating listItem={listItem} /> : null}
            </div>
            <div css={css({marginLeft: 10})}>
              <div
                css={css({
                  marginTop: '0.4em',
                  fontStyle: 'italic',
                  fontSize: '0.85em',
                })}
              >
                {author}
              </div>
              <small>{book.publisher}</small>
            </div>
          </div>
          <small>{book.synopsis.substring(0, 500)}...</small>
        </div>
      </Link>
      <div
        css={css({
          marginLeft: '20px',
          position: 'absolute',
          color: theme && theme.colors.gray80,
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'space-around',
          height: '100%',
        })}
      >
        <StatusButtons book={book} />
      </div>
    </div>
  )
}

export {BookRow}
