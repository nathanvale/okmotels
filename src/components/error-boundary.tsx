/* eslint-disable no-console */
/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react'

class ErrorBoundary extends React.Component<any, {hasError: boolean}> {
  constructor(props: any) {
    super(props)
    this.state = {hasError: false}
  }

  static getDerivedStateFromError() {
    // Update state so the next render will show the fallback UI.
    return {hasError: true}
  }

  componentDidCatch(error: Error, info: any) {
    // You can also log the error to an error reporting service
    console.log(error, info)
  }

  render() {
    if (this.state.hasError) {
      // You can render any custom fallback UI
      return <h1>Something went wrong.</h1>
    }

    return this.props.children
  }
}

export {ErrorBoundary}
