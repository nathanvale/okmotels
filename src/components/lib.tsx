import styled, {css} from 'styled-components'
import React from 'react'
import {CubeGrid} from 'styled-loaders-react'
import useTimeout from 'react-use/lib/useTimeout'

export const Centered = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100vh;
`

export const CircleButton = styled.button`
  border-radius: 30px;
  padding: 0;
  width: 40px;
  height: 40px;
  line-height: 1;
  display: flex;
  align-items: center;
  justify-content: center;
  background: ${props => props.theme.colors.base};
  color: ${props => props.theme.colors.text};
  border: 1px solid ${props => props.theme.colors.gray10};
  cursor: pointer;
`

export const BookListUL = styled.ul`
  list-style: none;
  padding: 0;
  display: grid;
  grid-template-rows: repeat(auto-fill, minmax(100px, 1fr));
  grid-gap: 1em;
`

export const Button = styled.button<{
  variant?: 'primary' | 'secondary'
}>`
  padding: 10px 15px;
  border: 0;
  line-height: 1;
  ${({variant = 'primary', theme}) =>
    variant === 'primary'
      ? `background: ${theme.colors.indigo}
        color: ${theme.colors.base2}`
      : `background: ${theme.colors.gray}
        color: ${theme.colors.text}`};
`

export const FormGroup = styled.div`
  display: flex;
  flex-direction: column;
`

export const Spinner = () => (
  <div aria-label="loading">
    <CubeGrid />
  </div>
)

export function FullPageSpinner() {
  const ready = useTimeout(0)
  return (
    <div css={css({marginTop: '3em', fontSize: '4em'})}>
      {ready ? <Spinner /> : <Spinner />}
    </div>
  )
}
