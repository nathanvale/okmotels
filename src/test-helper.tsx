import React from 'react'

import {render, RenderResult, RenderOptions} from 'react-testing-library'
import {ThemeProvider} from 'styled-components'
import {Omit} from './helper'

export const mock = <T extends {}, K extends keyof T>(
  object: T,
  property: K,
  value: T[K],
): void => {
  Object.defineProperty(object, property, {get: () => value})
}

export const renderWithTheme = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  theme: any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ui: React.ReactElement<any>,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult =>
  render(<ThemeProvider theme={theme}>{ui}</ThemeProvider>, options)
