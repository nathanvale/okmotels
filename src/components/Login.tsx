import React from 'react'
import {navigate} from 'gatsby'
import {RouteComponentProps} from '@reach/router'
import Helmet from 'react-helmet'
import {handleLogin, isLoggedIn} from '../services/auth'
import {Form, Submit, Input} from '../components/FormElements'
import {PageTitle} from '../components/PageTitle'
import {Container} from '../components/Container'

class Login extends React.Component<RouteComponentProps> {
  state = {
    username: ``,
    password: ``,
  }

  handleUpdate = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({
      [event.target.name]: event.target.value,
    })
  }

  handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault()
    handleLogin(this.state)
  }

  render(): JSX.Element {
    if (isLoggedIn()) {
      navigate(`/app/profile`)
    }

    return (
      <>
        <Helmet>
          <title>{`Login`}</title>
        </Helmet>
        <Container data-testid="login">
          <PageTitle>Login</PageTitle>
          <Form
            method="post"
            onSubmit={event => {
              this.handleSubmit(event)
              navigate(`/app/profile`)
            }}
          >
            <label hidden htmlFor="username">
              Username
            </label>
            <Input
              type="text"
              id="username"
              name="username"
              placeholder="Username"
              onChange={this.handleUpdate}
            />
            <label hidden htmlFor="password">
              Password
            </label>
            <Input
              type="password"
              id="password"
              name="password"
              placeholder="Password"
              onChange={this.handleUpdate}
            />
            <Submit type="submit" value="Log In" />
          </Form>
        </Container>
      </>
    )
  }
}

export {Login}
