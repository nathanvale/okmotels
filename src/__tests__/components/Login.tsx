/* import React from 'react'
import {axe} from 'jest-axe'
import {Login} from '../../components/Login'
import defaultTheme from '../../styles/theme'
import {renderWithTheme} from '../../test-helper' */

test('should have no accessibilty violations ', async () => {
  /*   const {container} = renderWithTheme(defaultTheme, <Login />)
  expect(await axe(container.innerHTML)).toHaveNoViolations()
  expect(container).toMatchSnapshot() */
})
