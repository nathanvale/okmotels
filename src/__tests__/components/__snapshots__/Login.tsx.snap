// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have no accessibilty violations  1`] = `
<div>
  <section
    class="Container__Wrapper-koxfXz bHtbAd"
    data-testid="login"
  >
    <h1
      class="PageTitle__Title-euxURV bAxgis"
    >
      Login
    </h1>
    <form
      class="FormElements__Form-gUrTuf cFQswl"
      method="post"
    >
      <label
        for="username"
        hidden=""
      >
        Username
      </label>
      <input
        class="FormElements__Input-kwjIXe extkka"
        id="username"
        name="username"
        placeholder="Username"
        type="text"
      />
      <label
        for="password"
        hidden=""
      >
        Password
      </label>
      <input
        class="FormElements__Input-kwjIXe extkka"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
      />
      <input
        class="FormElements__Submit-csJEoT brxrRQ"
        type="submit"
        value="Log In"
      />
    </form>
  </section>
</div>
`;
